File: src/main/kotlin/com/mobaassistant/service/CombatAnalysisService.kt
package com.mobaassistant.service

import com.mobaassistant.model.Hero
import org.springframework.stereotype.Service
import kotlin.math.max

@Service
class CombatAnalysisService {
    
    fun calculateDamageOutput(
        attacker: Hero,
        target: Hero,
        items: List<String>
    ): CombatPrediction {
        val baseDamage = calculateBaseDamage(attacker, items)
        val defense = calculateDefense(target, items)
        val penetration = calculatePenetration(attacker, items)
        
        val effectiveDamage = applyDamageFormula(baseDamage, defense, penetration)
        val timeToKill = calculateTimeToKill(effectiveDamage, target.stats.durability)
        
        return CombatPrediction(
            damagePerSecond = effectiveDamage,
            timeToKill = timeToKill,
            winProbability = calculateWinProbability(attacker, target, effectiveDamage),
            tradingAdvice = getTradingAdvice(attacker, target, effectiveDamage)
        )
    }

    fun getOptimalSkillCombos(hero: Hero, situation: String): List<SkillCombo> {
        return when (situation.lowercase()) {
            "burst" -> listOf(
                SkillCombo(
                    sequence = listOf("ultimate", "skill2", "skill1"),
                    damage = calculateComboDamage(hero, listOf("ultimate", "skill2", "skill1")),
                    difficulty = "Medium",
                    timing = "Use when enemy has no escape skills"
                )
            )
            "poke" -> listOf(
                SkillCombo(
                    sequence = listOf("skill1", "basic_attack"),
                    damage = calculateComboDamage(hero, listOf("skill1", "basic_attack")),
                    difficulty = "Easy",
                    timing = "Use to harass in lane"
                )
            )
            else -> emptyList()
        }
    }

    private fun calculateBaseDamage(hero: Hero, items: List<String>): Double {
        // Implement actual damage calculation logic
        return 100.0
    }

    private fun calculateDefense(hero: Hero, items: List<String>): Double {
        // Implement defense calculation logic
        return 50.0
    }

    private fun calculatePenetration(hero: Hero, items: List<String>): Double {
        // Implement penetration calculation logic
        return 30.0
    }

    private fun applyDamageFormula(damage: Double, defense: Double, penetration: Double): Double {
        return damage * (1 - (defense - penetration) / 100)
    }

    private fun calculateTimeToKill(dps: Double, hp: Int): Double {
        return hp / dps
    }

    private fun calculateWinProbability(
        attacker: Hero,
        target: Hero,
        effectiveDamage: Double
    ): Double {
        // Implement win probability calculation
        return 0.65
    }

    private fun getTradingAdvice(
        attacker: Hero,
        target: Hero,
        effectiveDamage: Double
    ): String {
        return "Favorable trade - engage when skill 2 is available"
    }

    private fun calculateComboDamage(hero: Hero, skills: List<String>): Double {
        // Implement combo damage calculation
        return 500.0
    }
}

data class CombatPrediction(
    val damagePerSecond: Double,
    val timeToKill: Double,
    val winProbability: Double,
    val tradingAdvice: String
)

data class SkillCombo(
    val sequence: List<String>,
    val damage: Double,
    val difficulty: String,
    val timing: String
)

File: app/services/combat-analysis.service.ts
import { Observable } from '@nativescript/core';
import { Hero } from '../models/hero.model';
import { GameState } from '../models/game-state.model';
import { CombatPrediction, SkillCombo } from '../models/combat.model';

export class CombatAnalysisService extends Observable {
  private gameState: GameState | null = null;

  calculateDamageOutput(
    attacker: Hero,
    target: Hero,
    items: string[]
  ): CombatPrediction {
    const baseDamage = this.calculateBaseDamage(attacker, items);
    const defense = this.calculateDefense(target, items);
    const penetration = this.calculatePenetration(attacker, items);
    
    const effectiveDamage = this.applyDamageFormula(baseDamage, defense, penetration);
    const timeToKill = this.calculateTimeToKill(effectiveDamage, target.stats.hp);
    
    return {
      damagePerSecond: effectiveDamage,
      timeToKill,
      winProbability: this.calculateWinProbability(attacker, target, effectiveDamage),
      tradingAdvice: this.getTradingAdvice(attacker, target, effectiveDamage)
    };
  }

  getOptimalSkillCombos(hero: Hero, situation: string): SkillCombo[] {
    const combos: SkillCombo[] = [];
    
    switch (situation) {
      case 'burst':
        combos.push({
          sequence: ['ultimate', 'skill2', 'skill1'],
          damage: this.calculateComboDamage(hero, ['ultimate', 'skill2', 'skill1']),
          difficulty: 'Medium',
          timing: 'Use when enemy has no escape skills'
        });
        break;
      case 'poke':
        combos.push({
          sequence: ['skill1', 'basic_attack'],
          damage: this.calculateComboDamage(hero, ['skill1', 'basic_attack']),
          difficulty: 'Easy',
          timing: 'Use to harass in lane'
        });
        break;
      case 'escape':
        combos.push({
          sequence: ['skill2', 'sprint'],
          damage: 0,
          difficulty: 'Easy',
          timing: 'Use when ganked'
        });
        break;
    }

    return combos;
  }

  getPositioningAdvice(
    hero: Hero,
    teamPosition: number[],
    enemyPosition: number[]
  ): string[] {
    const advice: string[] = [];
    const role = hero.role.toLowerCase();

    // Role-specific positioning
    switch (role) {
      case 'marksman':
        advice.push('Stay behind tanks, maintain maximum attack range');
        advice.push('Position near walls for flicker escape');
        break;
      case 'tank':
        advice.push('Position between enemy and allies');
        advice.push('Block skill shots for carries');
        break;
      case 'assassin':
        advice.push('Hide in bushes, wait for enemy skills to be used');
        advice.push('Position for backline access');
        break;
    }

    // Situation-specific advice
    if (this.isTeamfightImminent()) {
      advice.push('Group with team, maintain formation');
    }
    if (this.isObjectiveActive()) {
      advice.push('Control objective zone, zone out enemies');
    }

    return advice;
  }

  calculateTeamfightWinProbability(): number {
    if (!this.gameState) return 0.5;

    const factors = {
      goldDifference: this.calculateGoldAdvantage(),
      levelAdvantage: this.calculateLevelAdvantage(),
      teamComposition: this.evaluateTeamComposition(),
      objectiveControl: this.evaluateObjectiveControl()
    };

    // Weighted calculation of win probability
    return (
      factors.goldDifference * 0.3 +
      factors.levelAdvantage * 0.25 +
      factors.teamComposition * 0.25 +
      factors.objectiveControl * 0.2
    );
  }

  private calculateBaseDamage(hero: Hero, items: string[]): number {
    // Implementation of damage calculation
    return 100; // Simplified
  }

  private calculateDefense(hero: Hero, items: string[]): number {
    // Implementation of defense calculation
    return 50; // Simplified
  }

  private calculatePenetration(hero: Hero, items: string[]): number {
    // Implementation of penetration calculation
    return 30; // Simplified
  }

  private applyDamageFormula(
    damage: number,
    defense: number,
    penetration: number
  ): number {
    // Actual damage formula implementation
    return (damage * (1 - (defense - penetration) / 100));
  }

  private calculateTimeToKill(dps: number, hp: number): number {
    return hp / dps;
  }

  private calculateWinProbability(
    attacker: Hero,
    target: Hero,
    effectiveDamage: number
  ): number {
    // Complex win probability calculation
    return 0.65; // Simplified
  }

  private getTradingAdvice(
    attacker: Hero,
    target: Hero,
    effectiveDamage: number
  ): string {
    // Generate trading advice based on damage calculations
    return 'Favorable trade - engage when skill 2 is available';
  }

  private calculateComboSequence(hero: Hero): string[] {
    // Calculate optimal skill sequence
    return ['skill1', 'skill2', 'ultimate'];
  }

  private calculateComboTiming(combo: string[]): string {
    // Calculate best timing for combo execution
    return 'Execute when enemy dash is on cooldown';
  }

  private isTeamfightImminent(): boolean {
    // Logic to detect imminent teamfights
    return true;
  }

  private isObjectiveActive(): boolean {
    // Check if major objective is active
    return false;
  }

  private calculateGoldAdvantage(): number {
    if (!this.gameState) return 0;
    
    const allyGold = this.gameState.allyTeam.totalGold;
    const enemyGold = this.gameState.enemyTeam.totalGold;
    
    return (allyGold - enemyGold) / (allyGold + enemyGold);
  }

  private calculateLevelAdvantage(): number {
    if (!this.gameState) return 0;
    
    const allyLevels = this.gameState.allyTeam.players.reduce(
      (sum, player) => sum + player.level, 0
    );
    const enemyLevels = this.gameState.enemyTeam.players.reduce(
      (sum, player) => sum + player.level, 0
    );
    
    return (allyLevels - enemyLevels) / 25; // Normalized to -1 to 1
  }

  private evaluateTeamComposition(): number {
    // Complex team composition evaluation
    return 0.7;
  }

  private evaluateObjectiveControl(): number {
    // Evaluate objective control status
    return 0.6;
  }

  updateGameState(newState: GameState): void {
    this.gameState = newState;
    this.notifyPropertyChange('gameState', newState);
  }
}